compiler/packages/babel-plugin-react-compiler/src/HIR/PruneNonEscapingScopesHIR.ts

```diff
@@ -0,0 +1,957 @@
+      CompilerError.invariant(false, {
+        reason: `Unexpected unsupported node`,
+        description: null,
+        loc: value.loc,
+        suggestions: null,
+      });
+    }
+    default: {
+      assertExhaustive(
+        value,
+        `Unexpected value kind \`${(value as any).kind}\``,
+      );
+    }
+  }
+}
+
+function computePatternLValues(pattern: Pattern): Array<LValueMemoization> {
+  const lvalues: Array<LValueMemoization> = [];
+  switch (pattern.kind) {
+    case 'ArrayPattern': {
+      for (const item of pattern.items) {
+        if (item.kind === 'Identifier') {
+          lvalues.push({place: item, level: MemoizationLevel.Conditional});
+        } else if (item.kind === 'Spread') {
+          lvalues.push({place: item.place, level: MemoizationLevel.Memoized});
+        }
+      }
+      break;
+    }
+    case 'ObjectPattern': {
+      for (const property of pattern.properties) {
+        if (property.kind === 'ObjectProperty') {
+          lvalues.push({
+            place: property.place,
+            level: MemoizationLevel.Conditional,
+          });
+        } else {
+          lvalues.push({
+            place: property.place,
+            level: MemoizationLevel.Memoized,
+          });
+        }
+      }
+      break;
+    }
+    default: {
+      assertExhaustive(
+        pattern,
+        `Unexpected pattern kind \`${(pattern as any).kind}\``,
+      );
+    }
+  }
+  return lvalues;
+}
+
+/*
+ * Populates the input state with the set of returned identifiers and information about each
+ * identifier's and scope's dependencies.
+ */
+function collectDependencies(fn: HIRFunction, state: State): void {
+  let changed;
+  let undefinedIdentifiers: null | (() => Array<Place>);
+  do {
+    changed = false;
+    undefinedIdentifiers = null;
+    for (const [, block] of fn.body.blocks) {
+      for (const phi of block.phis) {
+        /*
+         * Need to be able to handle phis with undefined identifiers due to backpointers, but they should
+         * always become defined by the time we exit the fixpoint
+         */
+        const rvaluesUnfiltered = Array.from(phi.operands.values());
+        const rvalues = rvaluesUnfiltered.filter(
+          place =>
+            !state.inScope(
+              block.instructions[0]?.id ?? block.terminal,
+              place,
+            ) || state.identifiers.has(place.identifier.id),
+        );
+        const aliasing: Aliasing = {
+          lvalues: [{place: phi.place, level: MemoizationLevel.Conditional}],
+          rvalues,
+        };
+
+        /*
+         * Annoyingly complicated logic, but ultimately just so we can emit a useful error message if the
+         * invariants are violated
+         */
+        if (rvalues.length !== rvaluesUnfiltered.length) {
+          undefinedIdentifiers = (): Array<Place> =>
+            (undefinedIdentifiers?.() ?? []).concat(
+              rvaluesUnfiltered.filter(v => new Set(rvalues).has(v)),
```

Comment:
This is O(n^2) since it's creating a new Set for each instance, let's cache the Set. Also, i'm curious why make `undefinedIdentifiers` a function rather than an array or set of places?